source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,600 @@
+
+    // We need to make sure that OpSampledImage instructions will not be
+    // separated from their use, as they need to be in the same block.
+
+    // All result ids of an OpSampledImage instruction occurring before the last
+    // point where the block will need to be split.
+    std::set<uint32_t> sampled_image_result_ids_before_split;
+
+    // All result ids of an OpSampledImage instruction occurring after the last
+    // point where the block will need to be split. They can still be used.
+    std::set<uint32_t> sampled_image_result_ids_after_split;
+
+    // Check all of the instructions in the block.
+    bool all_instructions_compatible = block->WhileEachInst(
+        [this, &ir_context, &instructions_to_fresh_ids, &used_fresh_ids,
+         &overflow_ids_used, &sampled_image_result_ids_before_split,
+         &sampled_image_result_ids_after_split](opt::Instruction* instruction) {
+          // The instruction cannot be an atomic or barrier instruction
+          if (instruction->IsAtomicOp() ||
+              instruction->opcode() == SpvOpControlBarrier ||
+              instruction->opcode() == SpvOpMemoryBarrier ||
+              instruction->opcode() == SpvOpNamedBarrierInitialize ||
+              instruction->opcode() == SpvOpMemoryNamedBarrier ||
+              instruction->opcode() == SpvOpTypeNamedBarrier) {
+            return false;
+          }
+
+          // If the instruction is OpSampledImage, add the result id to
+          // |sampled_image_result_ids_after_split|.
+          if (instruction->opcode() == SpvOpSampledImage) {
+            sampled_image_result_ids_after_split.emplace(
+                instruction->result_id());
+          }
+
+          // If the instruction uses an OpSampledImage that appeared before a
+          // point where we need to split the block (before a load, store or
+          // function call), then the transformation is not applicable.
+          if (!instruction->WhileEachInId(
+                  [&sampled_image_result_ids_before_split](
+                      uint32_t* id) -> bool {
+                    return !sampled_image_result_ids_before_split.count(*id);
+                  })) {
+            return false;
+          }
+
+          // If the instruction is a load, store or function call, there must
+          // be a mapping from the corresponding instruction descriptor to a
+          // list of fresh ids or there must be enough overflow ids.
+          if (instruction->opcode() == SpvOpLoad ||
+              instruction->opcode() == SpvOpStore ||
+              instruction->opcode() == SpvOpFunctionCall) {
+            // Keep a vector of the fresh ids needed by this instruction.
+            std::vector<uint32_t> fresh_ids;
+            uint32_t overflow_ids_needed;
+
+            // Initialise overflow_ids_needed to the total number of ids
+            // needed.
+            overflow_ids_needed =
+                NumOfFreshIdsNeededByOpcode(instruction->opcode());
```

Comment:
The code for computing a precise number of fresh ids used by each memory instruction is smart but I find it hard to understand. 

It might be better to simply have a counter and increment it every time we find a memory instruction. If the counter is greater than the size of the `overflow_ids` - return false.

Then, as a separate check, we could ensure that all ids in `overflow_ids` are unique and fresh. We might check more ids than we will actually use in the transformation but I don't think this is a problem. @paulthomson, what do you think regarding this check causing any problems in the shrinker?

Reply:
Yes, I agree that it was a bit convoluted so I changed it so that the inner loop only checks if there are enough fresh ids, and the check that they are distinct happens outside.